federatedscope/vertical_fl/model/Tree.py

```diff
@@ -112,6 +114,88 @@ def update_child(self, node_num, left_child, right_child):
                   2].indicator = self.tree[node_num].indicator * right_child
 
 
+class DecisionTree(Tree):
+    def __init__(self, max_depth, lambda_, gamma):
+        super().__init__(max_depth, lambda_, gamma)
+        self.task_type = None  # ['classification', 'regression']
+
+    def _gini(self, indicator, y):
+        total_num = np.sum(indicator)
+        positive_num = np.dot(indicator, y)
+        negative_num = total_num - positive_num
+        return 1 - (positive_num / total_num)**2 - (negative_num /
+                                                    total_num)**2
+
+    def _check_same_label(self, y, indicator):
+        active_idx = np.nonzero(indicator)[0]
+        active_y = y[active_idx]
+        if np.sum(active_y) in [0, len(active_y)]:
+            return True
+        return False
+
+    def cal_gini(self, split_idx, y, indicator):
+        if self._check_same_label(y, indicator):
+            return np.float('inf')
```

Comment:
I have reconsidered about the def of gini, here return 1 may be more reasonable, since the maximal value of gini is 1.